core/src/main/java/io/questdb/Telemetry.java

```diff
@@ -24,17 +24,137 @@
-    public static final short SYSTEM_ILP_RESERVE_WRITER = 102;
+import io.questdb.cairo.*;
+import io.questdb.cairo.security.AllowAllCairoSecurityContext;
+import io.questdb.griffin.SqlCompiler;
+import io.questdb.griffin.SqlException;
+import io.questdb.griffin.SqlExecutionContext;
+import io.questdb.log.Log;
+import io.questdb.log.LogFactory;
+import io.questdb.mp.MPSequence;
+import io.questdb.mp.QueueConsumer;
+import io.questdb.mp.RingQueue;
+import io.questdb.mp.SCSequence;
+import io.questdb.std.Misc;
+import io.questdb.std.ObjectFactory;
+import io.questdb.std.Os;
+import io.questdb.std.datetime.microtime.MicrosecondClock;
+import io.questdb.tasks.AbstractTelemetryTask;
+import org.jetbrains.annotations.NotNull;
+
+import java.io.Closeable;
+
+public final class Telemetry<T extends AbstractTelemetryTask> implements Closeable {
+    private static final Log LOG = LogFactory.getLog(Telemetry.class);
+
+    private final boolean enabled;
+    private MicrosecondClock clock;
+    private MPSequence telemetryPubSeq;
+    private RingQueue<T> telemetryQueue;
+    private SCSequence telemetrySubSeq;
+    private TelemetryType<T> telemetryType;
+    private TableWriter writer;
+
+    private final QueueConsumer<T> taskConsumer = this::consume;
+
+    public Telemetry(TelemetryType<T> type, CairoConfiguration configuration) {
+        final TelemetryConfiguration telemetryConfiguration = type.getTelemetryConfiguration(configuration);
+        enabled = telemetryConfiguration.getEnabled();
+        if (enabled) {
+            clock = configuration.getMicrosecondClock();
+            telemetryType = type;
+            telemetryQueue = new RingQueue<>(type.getTaskFactory(), telemetryConfiguration.getQueueCapacity());
+            telemetryPubSeq = new MPSequence(telemetryQueue.getCycle());
+            telemetrySubSeq = new SCSequence();
+            telemetryPubSeq.then(telemetrySubSeq).then(telemetryPubSeq);
+        }
+    }
+
+    @Override
+    public void close() {
+        if (writer == null) {
+            return;
+        }
+
+        consumeAll();
+        Misc.free(telemetryQueue);
+
+        telemetryType.logStatus(writer, TelemetrySystemEvent.SYSTEM_DOWN, clock.getTicks());
+        writer = Misc.free(writer);
+    }
+
+    public void consume(T task) {
+        task.writeTo(writer);
+    }
+
+    public void consumeAll() {
+        if (enabled && telemetrySubSeq.consumeAll(telemetryQueue, taskConsumer)) {
+            writer.commit();
+        }
+    }
+
+    public void init(CairoEngine engine, SqlCompiler compiler, SqlExecutionContext sqlExecutionContext) throws SqlException {
+        if (!enabled) {
+            return;
+        }
+
+        compiler.compile(telemetryType.getCreateSql(), sqlExecutionContext);
+
+        final TableToken tableToken = engine.getTableToken(telemetryType.getTableName());
+        try {
+            writer = engine.getWriter(AllowAllCairoSecurityContext.INSTANCE, tableToken, "telemetry");
```

Comment:
nice to have: would be super useful if we can override column buffers to something very small here, like 1Mb instead of defaults so that the table takes fewer disk space / mapped memory.

Reply:
good idea, will change it.
could go into telemetry config.

Reply:
Currently, the column memory logic resides deep inside the `TableWriter`. Should this refactoring be placed in this PR?

Reply:
We can have alternative memory buffer sizes for system tables. This table has to be recognised as a system table so there is a way to use alternative configuration.
It's nice to have, if the code is getting too ugly we can forego it.

Reply:
I would prefer to ignore this optimisation for now and implement it as a separate PR.